### PR TITLE
[WIP] From 4 to 2 domains, and adding masker functionality

### PR DESCRIPTION
**Pull request by 51N84D**
Instead of four domains {real, sim}_{flooded, nonflooded} we now only have two {real, sim}. I've also added a "Masker" decoder. I've retained the "Water" decoder because we may want to use that for the auxiliary task of water segmentation (which is different from mask generation). 

Also, all of the tests should pass but I haven't tried running a real training experiment.

### CODE REVIEW

File: omnigan/data.py [hunk 1]
```diff
@@ -81,6 +81,11 @@ def pil_image_loader(path, task):
         arr = arr.astype(np.float32)
         arr[arr != 0] = 1 / arr[arr != 0]
 
+    if task == 'm':
+        arr[arr != 0] = 1 / arr[arr != 0]
+        #Make sure mask is RGB for the sake of transforms
```
Comment (vict0rsch) [inline omnigan/data.py:86]: depth loads inverse values, masks don't

File: omnigan/data.py [hunk 2]
```diff
@@ -81,6 +81,11 @@ def pil_image_loader(path, task):
...
+        if len(arr.shape) >= 3:
+            arr = arr[:,:,0]
```
Comment (vict0rsch) [inline omnigan/data.py:88]: how does that ensure rgb?

File: omnigan/data.py [hunk 3]
```diff
@@ -183,10 +189,10 @@ def check_samples(self):
                 assert Path(v).exists(), f"{k} {v} does not exist"
 
 
-def get_loader(domain, mode, opts):
+def get_loader(mode, domain, opts):
```
Comment (vict0rsch) [inline omnigan/data.py:192]: do you have an OCD? You didn't seem like that kind of guy

File: omnigan/generator.py [hunk 4]
```diff
@@ -87,6 +84,9 @@ def __init__(self, opts, latent_shape=None, verbose=None):
         if "w" in opts.tasks and not opts.gen.w.ignore:
             self.decoders["w"] = WaterDecoder(opts)
 
+        if "m" in opts.tasks and not opts.gen.m.ignore:
```
Comment (vict0rsch) [inline omnigan/generator.py:87]: I think you could remove "w" altogether from everywhere if `m` is going to do that job

File: omnigan/generator.py [hunk 5]
```diff
@@ -255,6 +238,19 @@ def __init__(self, opts):
         )
 
 
+class MaskDecoder(BaseDecoder):
+    def __init__(self, opts):
+        super().__init__(
+            opts.gen.w.n_upsample,
+            opts.gen.w.n_res,
```
Comment (vict0rsch) [inline omnigan/generator.py:245]: careful: trailing `.w.`

File: omnigan/trainer.py [hunk 6]
```diff
@@ -324,15 +315,30 @@ def run_epoch(self):
...
             # and send it to self.device
             print(
-                "\rEpoch {} batch {} step {}".format(
-                    self.logger.epoch, i, self.logger.global_step
-                )
+                "\rEpoch {} batch {} step {}".format(self.logger.epoch, i, self.logger.global_step)
             )
+
+            """
+            print("**********BATCH INFORMATION********************")
```
Comment (vict0rsch) [inline omnigan/trainer.py:322]: you and your strings as comments... :)

File: omnigan/trainer.py [hunk 7]
```diff
@@ -641,13 +652,9 @@ def get_translation_loss(self, multi_domain_batch):
...
             mask = (
-                torch.randint(0, 2, real_s_labels.shape)
-                .to(torch.float32)
-                .to(self.device)
+                torch.randint(0, 2, real_s_labels.shape).to(torch.float32).to(self.device)
             )  # TODO : load mask
-            update_loss = (
-                self.losses["G"]["t"]["sm"](fake_s, real_s_labels) * mask
-            ).mean()
+            update_loss = (self.losses["G"]["t"]["sm"](fake_s, real_s_labels) * mask).mean()
```
Comment (vict0rsch) [inline omnigan/trainer.py:657]: what does sm stand for?

Comment (vict0rsch) [reply omnigan/trainer.py:657]: let me rephrase: what does `sm` mean *in this context*?

File: omnigan/tutils.py [hunk 8]
```diff
@@ -92,17 +93,15 @@ def domains_to_class_tensor(domains, one_hot=False):
...
-        raise ValueError(
-            "Unknown domains {} should be in {}".format(domains, list(mapping.keys()))
-        )
+        raise ValueError("Unknown domains {} should be in {}".format(domains, list(mapping.keys())))
 
     target = torch.tensor([mapping[domain] for domain in domains])
 
     if one_hot:
-        one_hot_target = torch.FloatTensor(len(target), 4)  # 4 domains
+        one_hot_target = torch.FloatTensor(len(target), 2)  # 4 domains
```
Comment (vict0rsch) [inline omnigan/tutils.py:104]: edit comment too, at some point

File: omnigan/tutils.py [hunk 9]
```diff
@@ -304,7 +297,7 @@ def save_tanh_tensor(image, path):
     """
     path = Path(path)
     if isinstance(image, torch.Tensor):
-        image = image.detach().numpy()
+        image = image.detach().cpu().numpy()
```
Comment (vict0rsch) [inline omnigan/tutils.py:300]: good catch

**Comment by vict0rsch** [comment]
good wip :) 

updating tests is good writing new one is best :p (not immediately, but before merging)

File: omnigan/data.py (hunk 1 above)
Comment (51N84D) [reply omnigan/data.py:86]: True. I meant to binarize the masks here.

File: omnigan/data.py (hunk 2 above)
Comment (51N84D) [reply omnigan/data.py:88]: It does the opposite actually haha. I'll fix the comment

File: omnigan/data.py (hunk 3 above)
Comment (51N84D) [reply omnigan/data.py:192]: Lmao for the record, if you look at the code in master there's a mismatch:
`def get_loader(domain, mode, opts):
    return DataLoader(
        OmniListDataset(
            domain, mode, opts, transform=transforms.Compose(get_transforms(opts))
        ),
        batch_size=opts.data.loaders.get("batch_size", 4),
        # shuffle=opts.data.loaders.get("shuffle", True),
        shuffle=True,
        num_workers=opts.data.loaders.get("num_workers", 8),
    )


def get_all_loaders(opts):
    loaders = {}
    for mode in ["train", "val"]:
        loaders[mode] = {}
        for domain in ["rf", "rn", "sf", "sn"]:
            if mode in opts.data.files:
                if domain in opts.data.files[mode]:
                    loaders[mode][domain] = get_loader(mode, domain, opts)
    return loaders`

File: omnigan/generator.py (hunk 4 above)
Comment (51N84D) [reply omnigan/generator.py:87]: 'w' was meant to be a water classifier, not a mask-generator right? Though I suppose one could make the argument that our discriminators are basically water classifiers

File: omnigan/trainer.py (hunk 6 above)
Comment (51N84D) [reply omnigan/trainer.py:322]: For some reason this is a result of my auto-formatting. And yes, before you ask, I am using black ;)

File: omnigan/trainer.py (hunk 7 above)
Comment (51N84D) [reply omnigan/trainer.py:657]: No idea what it means. I've basically ignored everything to do with translation here so I missed this

File: omnigan/data.py (hunk 3 above)
Comment (vict0rsch) [reply omnigan/data.py:192]: haha ok my bad good catch :)

File: omnigan/generator.py (hunk 4 above)
Comment (vict0rsch) [reply omnigan/generator.py:87]: W was supposed to tell the model where to flood, but in the translation scheme that also means actual water segmentation when flood->non-flood so if we drop that it's just the masker

File: omnigan/trainer.py (hunk 6 above)
Comment (vict0rsch) [reply omnigan/trainer.py:322]: I know you are, I can see it I'm so proud! I don't care about strings as commends, I'm not asking you to change it but I'm not sure how common that is. I guess one argument that could be made is that unlike comments they do take a little memory. But who cares, we're doing deep learning

File: omnigan/trainer.py (hunk 7 above)
Comment (vict0rsch) [reply omnigan/trainer.py:657]: ouuuuh semantic matching loss

File: omnigan/tutils.py (hunk 8 above)
Comment (51N84D) [reply omnigan/tutils.py:104]: Tru

File: omnigan/generator.py (hunk 4 above)
Comment (51N84D) [reply omnigan/generator.py:87]: Ah. Ok then, farewell "WaterDecoder"

File: omnigan/trainer.py [hunk 10]
```diff
@@ -233,7 +240,7 @@ def setup(self):
         self.P = {"s": get_mega_model()}  # P => pseudo labeling models
 
         self.g_opt, self.g_scheduler = get_optimizer(self.G, self.opts.gen.opt)
-        self.d_opt, self.d_scheduler = get_optimizer(self.D, self.opts.dis.opt)
+        # self.d_opt, self.d_scheduler = get_optimizer(self.D, self.opts.dis.opt)
```
Comment (vict0rsch) [inline omnigan/trainer.py:243]: wait how do you plan on training discriminators?

File: omnigan/trainer.py [hunk 11]
```diff
@@ -287,8 +310,8 @@ def train_loaders(self):
     def update_learning_rates(self):
         if self.g_scheduler is not None:
             self.g_scheduler.step()
-        if self.d_scheduler is not None:
-            self.d_scheduler.step()
+        # if self.d_scheduler is not None:
```
Comment (vict0rsch) [inline omnigan/trainer.py:313]: idem

File: omnigan/trainer.py [hunk 12]
```diff
@@ -340,15 +363,78 @@ def run_epoch(self):
             }
 
             self.update_g(multi_domain_batch)
-            self.update_d(multi_domain_batch)
+            # self.update_d(multi_domain_batch)
```
Comment (vict0rsch) [inline omnigan/trainer.py:366]: idem

**Comment by 51N84D** [comment]
Ok, I added back the discriminators. Is it ready to merge into master?

**Comment by melisandeteng** [comment]
@51N84D I think it's ready enough for Tianyu and I to start working on it!

File: shared/trainer/defaults.yaml [hunk 13]
```diff
@@ -2,7 +2,7 @@ output_path: /network/tmp1/schmidtv/yb_runs/test_v1
 # -------------------
 # -----  Tasks  -----
 # -------------------
-tasks: [a, d, h, s, t, w]
+tasks: [a, d, h, s, t, w, m]
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:5]: can you remove all references to `w` and WaterDecoder?

**Comment by melisandeteng** [comment]
> question before we merge: do we consider discriminator(s) for the masked to be part of the advent PR or should that already be a flag here? it would ~ boil down to adding a key in the OmniDiscriminator class and a key like `generative:True|False` in `m`

That would be useful! instead of having to comment the discriminators in and out

File: shared/trainer/defaults.yaml (hunk 13 above)
Comment (51N84D) [reply shared/trainer/defaults.yaml]: Done (as far as I can tell)

**Comment by 51N84D** [comment]
> > question before we merge: do we consider discriminator(s) for the masked to be part of the advent PR or should that already be a flag here? it would ~ boil down to adding a key in the OmniDiscriminator class and a key like `generative:True|False` in `m`
> 
> That would be useful! instead of having to comment the discriminators in and out

Right now the code is setup such that if the "OmniDiscriminator" is empty (as in no parameters) it gets ignored. I think we can wait to add the ADVENT key with the ADVENT PR

**Comment by melisandeteng** [comment]
> > > question before we merge: do we consider discriminator(s) for the masked to be part of the advent PR or should that already be a flag here? it would ~ boil down to adding a key in the OmniDiscriminator class and a key like `generative:True|False` in `m`
> > 
> > 
> > That would be useful! instead of having to comment the discriminators in and out
> 
> Right now the code is setup such that if the "OmniDiscriminator" is empty (as in no parameters) it gets ignored. I think we can wait to add the ADVENT key with the ADVENT PR

Oh ok I hadn't noticed that. I'm wondering is there a way  to ignore default settings then ?

**Comment by vict0rsch** [comment]
what do you mean ignore default? what problem do you have?

**Comment by 51N84D** [comment]
@vict0rsch actually, I may have made the commit that un-comments the discriminator stuff after Méli already cloned the branch...
@melisandeteng can you try pulling those changes and running again? You shouldn't have to comment anything out

**Comment by melisandeteng** [comment]
> @vict0rsch actually, I may have made the commit that un-comments the discriminator stuff after Méli already cloned the branch...
> @melisandeteng can you try pulling those changes and running again? You shouldn't have to comment anything out

It works perfectly fine! for some reason when I first had pulled the uncommented version it would output an error message with "Key  "a" does not exist" in the tasks so I thought it was fetching the default parameters but somehow I don't have this issue anymore!
